common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -385,6 +446,26 @@ protected final boolean runAllTasksFrom(Queue<Runnable> taskQueue) {
         }
     }
 
+    /**
+     * What ever tasks are present in {@code taskQueue} when this method is invoked will be {@link Runnable#run()}.
+     * @param taskQueue the task queue to drain.
+     * @return {@code true} if at least {@link Runnable#run()} was called.
+     */
+    private boolean runExistingTasksFrom(Queue<Runnable> taskQueue) {
+        taskQueue.offer(BOOKEND_TASK);
```

Comment:
I think this would be simpler if you called `size()` up front, and just did a normal for loop until that many tasks were executed.  If taskQueue is full, you run the risk of getting an NPE later on.

Reply:
This is an unbounded MPSC queue and size computation requires an iteration. I think insertion/removal should be a preferred approach here. We can make the assumptions about unbounded queue more clear or be more defensive if that helps.

Reply:
Actually the queue is unbounded by default, but the user may specify a limit ... let me make this more robust.